src/distributions/mod.rs

```diff
@@ -32,45 +31,46 @@ pub mod gamma;
 pub mod normal;
 pub mod exponential;
 
-/// Types that can be used to create a random instance of `Support`.
-pub trait Sample<Support> {
-    /// Generate a random value of `Support`, using `rng` as the
-    /// source of randomness.
-    fn sample<R: Rng>(&mut self, rng: &mut R) -> Support;
-}
-
-/// `Sample`s that do not require keeping track of state.
+/// Type that can be used to create a random instance of `Output`.
 ///
 /// Since no state is recorded, each sample is (statistically)
 /// independent of all others, assuming the `Rng` used has this
 /// property.
-// FIXME maybe having this separate is overkill (the only reason is to
-// take &self rather than &mut self)? or maybe this should be the
-// trait called `Sample` and the other should be `DependentSample`.
-pub trait IndependentSample<Support>: Sample<Support> {
-    /// Generate a random value.
-    fn ind_sample<R: Rng>(&self, &mut R) -> Support;
+pub trait Distribution {
+    type Output;
+    /// Generate a random value of `Output`, using `rng` as the
+    /// source of randomness.
+    fn sample<R: Rng>(&self, rng: &mut R) -> <Self as Distribution>::Output;
 }
 
-/// A wrapper for generating types that implement `Rand` via the
-/// `Sample` & `IndependentSample` traits.
-pub struct RandSample<Sup> {
-    _marker: marker::PhantomData<fn() -> Sup>,
+/// A wrapper for generating types that implement `Distribution` via the
+/// `Rand` trait.
+pub struct RandDistribution<T> {
+    _marker: marker::PhantomData<fn() -> T>,
 }
 
-impl<Sup: Rand> Sample<Sup> for RandSample<Sup> {
-    fn sample<R: Rng>(&mut self, rng: &mut R) -> Sup { self.ind_sample(rng) }
-}
+impl<T: Rand> Distribution for RandDistribution<T> {
+    type Output = T;
 
-impl<Sup: Rand> IndependentSample<Sup> for RandSample<Sup> {
-    fn ind_sample<R: Rng>(&self, rng: &mut R) -> Sup {
+    fn sample<R: Rng>(&self, rng: &mut R) -> T {
         rng.gen()
     }
 }
 
-impl<Sup> RandSample<Sup> {
-    pub fn new() -> RandSample<Sup> {
-        RandSample { _marker: marker::PhantomData }
+impl<T> RandDistribution<T> {
+    pub fn new() -> RandDistribution<T> {
+        RandDistribution { _marker: marker::PhantomData }
+    }
+}
+
+pub struct ConstantDistribution<T>(T);
+
+impl<T: Clone> Distribution for ConstantDistribution<T> {
+    type Output = T;
+
+    fn sample<R: Rng>(&self, _: &mut R) -> T {
+        let ConstantDistribution(ref value) = *self;
```

Comment:
This can be `self.0.clone()`: the fields of tuple structs (and tuples) can be accessed positionally.